examples/reporting/parallel_report_download.py

```diff
@@ -0,0 +1,191 @@
+import multiprocessing
+import sys
+import time
+
+from google.ads.google_ads.client import GoogleAdsClient
+from google.ads.google_ads.errors import GoogleAdsException
+
+# Maximum number of processes to spawn.
+MAX_PROCESSES = multiprocessing.cpu_count()
+# Timeout between retries in seconds.
+BACKOFF_FACTOR = 5
+# Maximum number of retries for errors.
+MAX_RETRIES = 5
+
+
+def main(client, customer_ids):
+    """The main method that creates all necessary entities for the example.
+
+    Args:
+        client: an initialized GoogleAdsClient instance.
+        customer_ids: an array of client customer IDs.
+    """
+
+    # Define the GAQL query strings to run for each customer ID.
+    campaign_query = """
+        SELECT campaign.id, metrics.impressions, metrics.clicks
+        FROM campaign
+        WHERE segments.date DURING LAST_30_DAYS"""
+    ad_group_query = """
+        SELECT campaign.id, ad_group.id, metrics.impressions, metrics.clicks
+        FROM ad_group
+        WHERE segments.date DURING LAST_30_DAYS"""
+
+    inputs = _generate_inputs(
+        client, customer_ids, [campaign_query, ad_group_query]
+    )
+    with multiprocessing.Pool(MAX_PROCESSES) as pool:
+        # Call _issue_search_request on each input, parallelizing the work
+        # across processes in the pool.
+        results = pool.starmap(_issue_search_request, inputs)
+
+        # Partition our results into successful and failed results.
+        successes = []
+        failures = []
+        for res in results:
+            if res[0]:
+                successes.append(res[1])
+            else:
+                failures.append(res[1])
+
+        # Output results.
+        print(
+            f'Total successful results: {len(successes)}\n'
+            f'Total failed results: {len(failures)}\n'
+        )
+
+        print("Successes:") if len(successes) else None
+        for success in successes:
+            # success["results"] represents an array of result strings for one
+            # customer ID / query combination.
+            result_str = "\n".join(success["results"])
+            print(result_str)
+
+        print("Failures:") if len(failures) else None
+        for failure in failures:
+            ex = failure["exception"]
+            print(
+                f'Request with ID "{ex.request_id}" failed with status '
+                f'"{ex.error.code().name}" for customer_id '
+                f'{failure["customer_id"]} and query "{failure["query"]}" and '
+                'includes the following errors:'
+            )
+            for error in ex.failure.errors:
+                print(f'\tError with message "{error.message}".')
+                if error.location:
+                    for (
+                        field_path_element
+                    ) in error.location.field_path_elements:
+                        print(f'\t\tOn field: {field_path_element.field_name}')
+
+
+def _issue_search_request(client, customer_id, query):
```

Comment:
Curious -- is the client cloned (i.e. serialized / deserialized) in this case, or not? In other words, if a particular instance of _issue_search_request modifies the client, does it affect other instances of _issue_search_request? 

A comment to clarify the nature of the "thread safety" of `client` in this case would be most helpful.

Reply:
Tested this -- if an instance of _issue_search_request modifies the client, it does not affect other instances. I think this is a case of protection by python's call-by-value design for immutable arguments (if there were setters on the client class, that'd be a different story).

Reply:
The workers are running in separate processes, not separate threads, so they can't interfere with each other since there is no shared memory between the processes.